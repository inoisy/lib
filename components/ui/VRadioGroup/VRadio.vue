<script>
/* eslint-disable no-empty-function,no-unused-vars,vue/require-prop-types,no-prototype-builtins,new-cap,vue/require-default-prop,no-unused-expressions */

// Components
import VLabel from '../VLabel';
import VIcon from '../VIcon';
import VInput from '../VInput';

// Mixins
import BindsAttrs from '../../../mixins/binds-attrs';
import Colorable from '../../../mixins/colorable';
import { factory as GroupableFactory } from '../../../mixins/groupable';
import Rippleable from '../../../mixins/rippleable';
import Themeable from '../../../mixins/themeable';
import Selectable, { prevent } from '../../../mixins/selectable';

// Utilities
import { getSlot } from '../../../utils/helpers';

import mixins from '../../../utils/mixins';
import { mergeListeners } from '../../../utils/mergeData';

const baseMixins = mixins(BindsAttrs, Colorable, Rippleable, GroupableFactory('radioGroup'), Themeable);

export default baseMixins.extend({
    name: 'VRadio',

    inheritAttrs: false,

    props: {
        disabled: Boolean,
        id: String,
        label: String,
        name: String,
        offIcon: {
            type: String,
            default: '$radioOff',
        },
        onIcon: {
            type: String,
            default: '$radioOn',
        },
        readonly: Boolean,
        value: {
            default: null,
        },
    },

    data() {
        return {
            isFocused: false,
        };
    },

    computed: {
        classes() {
            return {
                'v-radio--is-disabled': this.isDisabled,
                'v-radio--is-focused': this.isFocused,
                ...this.themeClasses,
                ...this.groupClasses,
            };
        },
        computedColor() {
            return Selectable.options.computed.computedColor.call(this);
        },
        computedIcon() {
            return this.isActive
                ? this.onIcon
                : this.offIcon;
        },
        computedId() {
            return VInput.options.computed.computedId.call(this);
        },
        hasLabel: VInput.options.computed.hasLabel,
        hasState() {
            return (this.radioGroup || {}).hasState;
        },
        isDisabled() {
            return this.disabled ||
                Boolean(this.radioGroup) &&
                this.radioGroup.isDisabled;
        },
        isReadonly() {
            return this.readonly ||
                Boolean(this.radioGroup) &&
                this.radioGroup.isReadonly;
        },
        computedName() {
            if (this.name || !this.radioGroup) {
                return this.name;
            }

            return this.radioGroup.name || `radio-${this.radioGroup._uid}`;
        },
        rippleState() {
            return Selectable.options.computed.rippleState.call(this);
        },
        validationState() {
            return (this.radioGroup || {}).validationState || this.computedColor;
        },
    },

    methods: {
        genInput(args) {
            // We can't actually use the mixin directly because
            // it's made for standalone components, but its
            // genInput method is exactly what we need
            return Selectable.options.methods.genInput.call(this, 'radio', args);
        },
        genLabel() {
            if (!this.hasLabel) {
                return null;
            }

            return this.$createElement(VLabel, {
                on: {
                    // Label shouldn't cause the input to focus
                    click: prevent,
                },
                attrs: {
                    for: this.computedId,
                },
                props: {
                    color: this.validationState,
                    focused: this.hasState,
                },
            }, getSlot(this, 'label') || this.label);
        },
        genRadio() {
            const { title, ...radioAttrs } = this.attrs$;

            return this.$createElement('div', {
                staticClass: 'v-input--selection-controls__input',
            }, [
                this.$createElement(VIcon, this.setTextColor(this.validationState, {
                    props: {
                        dense: this.radioGroup && this.radioGroup.dense,
                    },
                }), this.computedIcon),
                this.genInput({
                    name: this.computedName,
                    value: this.value,
                    ...radioAttrs,
                }),
                this.genRipple(this.setTextColor(this.rippleState)),
            ]);
        },
        onFocus(e) {
            this.isFocused = true;
            this.$emit('focus', e);
        },
        onBlur(e) {
            this.isFocused = false;
            this.$emit('blur', e);
        },
        onChange() {
            if (this.isDisabled || this.isReadonly || this.isActive) {
                return;
            }

            this.toggle();
        },
        onKeydown: () => {}, // Override default with noop
    },

    render(h) {
        const data = {
            staticClass: 'v-radio',
            class: this.classes,
            on: mergeListeners({
                click: this.onChange,
            }, this.listeners$),
            attrs: { title: this.attrs$.title },
        };

        return h('div', data, [
            this.genRadio(),
            this.genLabel(),
        ]);
    },
});

</script>
<style lang="scss">
    @import './_variables';


    /* stylelint-disable declaration-no-important,no-duplicate-selectors */

    @include theme(v-radio) using ($material) {
        &--is-disabled {
            label {
                color: map-deep-get($material, 'text', 'disabled');
            }

            .v-icon {
                // needed for helper override
                color: map-deep-get($material, 'selection-controls', 'disabled') !important;
            }
        }
    }

    .v-radio {
        display: flex;
        align-items: center;
        height: auto;
        outline: none;

        &--is-disabled {
            cursor: default;
            pointer-events: none;
        }
    }

    .v-input--radio-group.v-input--radio-group--row {
        .v-radio {
            margin-right: $radio-margin-right;
        }
    }
</style>
