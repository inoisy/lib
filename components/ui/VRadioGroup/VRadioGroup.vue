<script>
/* eslint-disable no-empty-function,no-unused-vars,vue/require-prop-types,no-prototype-builtins,new-cap,vue/require-default-prop,no-unused-expressions */

// Extensions
import VInput from '../VInput';
import { BaseItemGroup } from '../VItemGroup/VItemGroup.vue';

// Mixins
import Comparable from '../../../mixins/comparable';

import mixins from '../../../utils/mixins';

const baseMixins = mixins(Comparable, BaseItemGroup, VInput);

export default baseMixins.extend({
    name: 'VRadioGroup',

    provide() {
        return {
            radioGroup: this,
        };
    },

    props: {
        column: {
            type: Boolean,
            default: true,
        },
        height: {
            type: [Number, String],
            default: 'auto',
        },
        name: String,
        row: Boolean,
        // If no value set on VRadio
        // will match valueComparator
        // force default to null
        value: {},
    },

    computed: {
        classes() {
            return {
                ...VInput.options.computed.classes.call(this),
                'v-input--selection-controls v-input--radio-group': true,
                'v-input--radio-group--column': this.column && !this.row,
                'v-input--radio-group--row': this.row,
            };
        },
    },

    methods: {
        genDefaultSlot() {
            return this.$createElement('div', {
                staticClass: 'v-input--radio-group__input',
                attrs: {
                    id: this.id,
                    role: 'radiogroup',
                    'aria-labelledby': this.computedId,
                },
            }, VInput.options.methods.genDefaultSlot.call(this));
        },
        genInputSlot() {
            const render = VInput.options.methods.genInputSlot.call(this);

            delete render.data?.on?.click;

            return render;
        },
        genLabel() {
            const label = VInput.options.methods.genLabel.call(this);

            if (!label) {
                return null;
            }
            if (label.data?.attrs) {
                label.data.attrs.id = this.computedId;
            }

            // WAI considers this an orphaned label
            if (label.data?.attrs) {
                delete label.data?.attrs?.for;
            }

            label.tag = 'legend';

            return label;
        },
        onClick: BaseItemGroup.options.methods.onClick,
    },
});

</script>
<style lang="scss">
    @import './_variables';
    @import '../../../styles/components/_selection-controls';

    .v-input--radio-group {
        legend.v-label {
            height: auto;
            font-size: 14px;
            cursor: text;
        }

        &__input {
            display: flex;
            width: 100%;
            border: none;
            cursor: default;
        }

        &--column .v-input--radio-group__input > .v-label {
            padding-bottom: $radio-group-padding;
        }

        &--row .v-input--radio-group__input > .v-label {
            padding-right: $radio-group-padding;
        }

        &--row {
            legend {
                display: inline-block;
                align-self: center;
            }

            .v-input--radio-group__input {
                flex-direction: row;
                flex-wrap: wrap;
            }
        }

        &--column {
            legend {
                padding-bottom: $radio-group-padding;
            }

            .v-radio:not(:last-child):not(:only-child) {
                margin-bottom: $radio-group-padding;
            }

            .v-input--radio-group__input {
                flex-direction: column;
            }
        }
    }
</style>
