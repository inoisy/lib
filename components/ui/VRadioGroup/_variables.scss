
$radio-margin-right: 16px !default;
$radio-group-padding: 8px !default;
