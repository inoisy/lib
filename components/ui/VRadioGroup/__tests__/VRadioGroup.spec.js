// Components
import VRadio from '../VRadio';
import VRadioGroup from '../VRadioGroup';

// Utilities
import {
    mount,
} from '@vue/test-utils';

describe('VRadioGroup.js', () => {
    let mountFunction;

    beforeEach(() => {
        mountFunction = (options = {}) => mount(VRadioGroup, options);
    });

    it('should match snapshot', async () => {
        const wrapper = mountFunction();

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('should match dense snapshot', async () => {
        const wrapper = mountFunction({
            propsData: {
                dense: true,
            },
            slots: {
                default: [VRadio],
            },
        });

        expect(wrapper.html()).toMatchSnapshot();
    });
});
