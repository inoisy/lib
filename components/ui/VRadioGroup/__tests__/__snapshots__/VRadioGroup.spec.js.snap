// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VRadioGroup.js should match dense snapshot 1`] = `
<div class="v-input v-input--dense theme--light v-input--selection-controls v-input--radio-group v-input--radio-group--column">
  <div class="v-input__control">
    <div class="v-input__slot"
         style="height: auto;"
    >
      <div role="radiogroup"
           aria-labelledby="input-5"
           class="v-input--radio-group__input"
      >
        <div class="v-radio theme--light">
          <div class="v-input--selection-controls__input">
            <span aria-hidden="true"
                  class="v-icon notranslate v-icon--dense theme--light"
            >
              <svg xmlns="http://www.w3.org/2000/svg"
                   viewbox="0 0 24 24"
                   role="img"
                   aria-hidden="true"
                   class="v-icon__svg"
              >
                <path d="M12,20C7.58,20 4,16.42 4,12C4,7.58 7.58,4 12,4C16.42,4 20,7.58 20,12C20,16.42 16.42,20 12,20M12,2C6.48,2 2,6.48 2,12C2,17.52 6.48,22 12,22C17.52,22 22,17.52 22,12C22,6.48 17.52,2 12,2Z">
                </path>
              </svg>
            </span>
            <input aria-checked="false"
                   id="input-6"
                   role="radio"
                   type="radio"
                   name="radio-5"
                   value
            >
            <div class="v-input--selection-controls__ripple">
            </div>
          </div>
        </div>
      </div>
    </div>
    <div class="v-messages theme--light">
      <transition-group-stub tag="div"
                             name="message-transition"
                             class="v-messages__wrapper"
      >
      </transition-group-stub>
    </div>
  </div>
</div>
`;

exports[`VRadioGroup.js should match snapshot 1`] = `
<div class="v-input theme--light v-input--selection-controls v-input--radio-group v-input--radio-group--column">
  <div class="v-input__control">
    <div class="v-input__slot"
         style="height: auto;"
    >
      <div role="radiogroup"
           aria-labelledby="input-1"
           class="v-input--radio-group__input"
      >
      </div>
    </div>
    <div class="v-messages theme--light">
      <transition-group-stub tag="div"
                             name="message-transition"
                             class="v-messages__wrapper"
      >
      </transition-group-stub>
    </div>
  </div>
</div>
`;
